File: Java/src/BinarySearchTree.java

public class BinarySearchTree {
	
	Node root;
	
	
	class Node {
		int value;
		Node left;
		Node right;
		
		Node(int value) {
			this.value = value;
		}
	}
	
	
	public boolean insert(int value) {
		Node node = new Node(value);
		Node temp = root;
		
		if (temp == null) {
			root = node;
			return true;
		}
		while (true) {
			if (value < temp.value) {
				if (temp.left == null) {
					temp.left = node;
					return true;
				} else {
					temp = temp.left;
				}
			} else if (value == temp.value) {
				return false;
			} else {
				if (temp.right == null) {
					temp.right = node;
					return true;
				} else {
					temp = temp.right;
				}
			}
		}
	}
	
	
	public boolean contains(int value) {
		Node temp = root;
		if (temp == null) return false;
		while (temp != null) {
			if (value < temp.value) {
				temp = temp.left;
			} else if (value == temp.value) {
				return true;
			} else {
				temp = temp.right;
			}
		}
		return false;
	}
}

File: Javascript/Extras/cheatsheet/BinarySearchTree.js
/*

NODE

Constructor
-- has a value and a left, right pointer --

------------

BINARY SEARCH TREE

Constructor
-- has a root node --

Methods
-- add, search, invert --

*/


class Node {

    constructor(value) {
        this.value = value;
        this.left = null;
        this.right = null;
    }

}

class BinarySearchTree {

    constructor() {
        this.root = null;
    }

    add = (value) => {
        const addedNode = new Node(value);
        if (this.root === null) {
            this.root = addedNode
            return
        }
        let iterator = this.root

        while (iterator !== null) {
            if (value === iterator.value) return
            if (value > iterator.value) {
                if (iterator.right === null) {
                    iterator.right = addedNode
                    return
                }
                iterator = iterator.right
            } else {
                if (iterator.left === null) {
                    iterator.left = addedNode
                    return
                }
                iterator = iterator.left
            }
        }
    }

    search = (value) => {
        let iterator = this.root
        while (iterator !== null) {
            if (value === iterator.value) return true
            if (value < iterator.value) {
                iterator = iterator.left
            } else {
                iterator = iterator.right
            }
        }
        return false
    }

    invert = (head = this.root) => {
        if (head === null) return
        let inverterNode = head.left
        head.left = head.right
        head.right = inverterNode
        this.invert(head.left)
        this.invert(head.right)

        return head
    }

}